packages/flutter/lib/src/material/bottom_navigation_bar.dart

```diff
@@ -0,0 +1,551 @@
+/// labels layed out on top of a piece of material. It provies quick navigation
+/// between top-level views of an app and is typically used on mobile. For
+/// larger screens, side navigation may be a better fit.
+///
+/// Bottom navigation bar is usually used in conjunction with [Scaffold] where
+/// it is provided as the [Scaffold.bottomNavigationBar] argument.
+///
+/// See also:
+///
+///  * [DestinationLabel]
+///  * [Scaffold]
+///  * <https://material.google.com/components/bottom-navigation.html>
+class BottomNavigationBar extends StatefulWidget {
+  /// Creates a bottom navigation bar, typically used in a [Scaffold] where it
+  /// is provided as the [Scaffold.bottomNavigationBar] argument.
+  ///
+  /// The arguments [labels] and [type] should not be null.
+  BottomNavigationBar({
+    @required this.labels,
+    this.onTap,
+    this.type: BottomNavigationBarType.fixed
+  }) {
+    assert(this.labels != null);
+    assert(this.type != null);
+  }
+
+  /// The interactive labels layed out within the bottom navigation bar.
+  final List<DestinationLabel> labels;
+
+  /// The callback that is called when a label is tapped.
+  final BottomNavigationBarCallback onTap;
+
+  /// The type of the bottom navigation bar that effects its layout.
+  final BottomNavigationBarType type;
+
+  @override
+  _BottomNavigationBarState createState() => new _BottomNavigationBarState();
+}
+
+class _BottomNavigationBarState extends State<BottomNavigationBar> {
+  int _currentIndex;
+  List<AnimationController> _controllers;
+  List<CurvedAnimation> _animations;
+  double _weight;
+  final Queue<_Circle> _circles = new Queue<_Circle>();
+  Color _backgroundColor;
+
+  static final Tween<double> _ratioTween = new Tween<double>(
+    begin: 1.0,
+    end: 1.5
+  );
+
+  @override
+  void initState() {
+    super.initState();
+    _controllers = new List<AnimationController>.generate(config.labels.length, (int index) {
+      return new AnimationController(
+        duration: const Duration(milliseconds: 200)
+      )..addListener(_rebuild);
+    });
+    _animations = new List<CurvedAnimation>.generate(config.labels.length, (int index) {
+      return new CurvedAnimation(
+        parent: _controllers[index],
+        curve: Curves.fastOutSlowIn,
+        reverseCurve: Curves.fastOutSlowIn.flipped
+      );
+    });
+
+    _currentIndex = 0;
+    _controllers[_currentIndex].value = 1.0;
+    if (config.onTap != null)
+      config.onTap(_animations[_currentIndex], _currentIndex);
+
+    _backgroundColor = config.labels[_currentIndex].backgroundColor;
+  }
+
+  @override
+  void dispose() {
+    for (AnimationController controller in _controllers)
+      controller.dispose();
+    for (_Circle circle in _circles)
+      circle.dispose();
+    super.dispose();
+  }
+
+  void _rebuild() {
+    setState(() {
+      // Animation tick.
+    });
+  }
+
+  double get _maxWidth {
+    assert(config.type != null);
```

Comment:
I was asserting here based on @Hixie's tip: this assertion here helps readability.